deep_quoridor/src/agents/flat_dqn.py

```diff
@@ -0,0 +1,193 @@
+
+        # Concatenate all components
+        flat_obs = np.concatenate([board, walls, my_walls, opponent_walls])
+        return torch.FloatTensor(flat_obs).to(self.device)
+
+    def get_action(self, game):
+        """
+        Select an action using epsilon-greedy policy.
+        """
+        observation, _, termination, truncation, _ = game.last()
+        if termination or truncation:
+            return None
+
+        mask = observation["action_mask"]
+        valid_actions = np.where(mask == 1)[0]
+
+        # With probability epsilon, select a random action (exploration)
+        if random.random() < self.epsilon:
+            return np.random.choice(valid_actions)
+
+        # Otherwise, select the action with the highest Q-value (exploitation)
+        state = self.preprocess_observation(observation)
+        with torch.no_grad():
+            q_values = self.online_network(state)
+
+        # Apply action mask to q_values
+        mask_tensor = torch.FloatTensor(mask).to(self.device)
+        q_values = q_values * mask_tensor - 1e9 * (1 - mask_tensor)
+
+        return torch.argmax(q_values).item()
+
+    def train(self, batch_size):
+        """
+        Train the network on a batch of samples from the replay buffer.
+        """
+        if len(self.replay_buffer) < batch_size:
+            return
+
+        # Sample a batch of transitions
+        batch = self.replay_buffer.sample(batch_size)
+        states, actions, rewards, next_states, dones = zip(*batch)
+
+        # Convert to tensors
+        states = torch.stack([torch.FloatTensor(s) for s in states]).to(self.device)
+        actions = torch.LongTensor(actions).unsqueeze(1).to(self.device)
+        rewards = torch.FloatTensor(rewards).to(self.device)
+        next_states = torch.stack([torch.FloatTensor(s) for s in next_states]).to(self.device)
+        dones = torch.FloatTensor(dones).to(self.device)
+
+        # Compute current Q values
+        current_q_values = self.online_network(states).gather(1, actions).squeeze()
+
+        # Compute next Q values using target network
+        with torch.no_grad():
+            next_q_values = self.target_network(next_states).max(1)[0]
+            target_q_values = rewards + (1 - dones) * self.gamma * next_q_values
+
+        # Compute loss and update online network
+        loss = self.criterion(current_q_values, target_q_values)
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+        # Decay epsilon
+        self.epsilon = max(self.epsilon_min, self.epsilon * self.epsilon_decay)
+
+        return loss.item()
+
+    def save_model(self, path):
+        """Save the model to disk."""
+        torch.save(self.online_network.state_dict(), path)
+
+    def load_model(self, path):
+        """Load the model from disk."""
+        self.online_network.load_state_dict(torch.load(path))
+        self.update_target_network()
+
+
+class Pretrained01FlatDQNAgent(FlatDQNAgent):
```

Comment:
This is a subclass that loads a specific model weights file. In the future this could be a parameter instead, so that we don't have to create dozens of subclasses per architecture.